std/format.d

```diff
@@ -7017,3 +7063,538 @@ char[] sformat(Char, Args...)(return scope char[] buf, scope const(Char)[] fmt,
     assert(c == "^-nan         $", "\ngot:'"~ c ~ "'\nexp:'^-nan         $'");
     assert(d == "^-NAN         $", "\ngot:'"~ d ~ "'\nexp:'^-NAN         $'");
 }
+
+private auto printFloat(T, Char)(T val, FormatSpec!Char f, FloatingPointControl.RoundingMode rm)
+if (is(T == float) || is(T == double) || (is(T == real) && T.mant_dig == double.mant_dig))
+{
+    import std.conv : to;
+    import std.algorithm.comparison : max;
+
+    enum byte[16] alpha = ['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'];
+    enum byte[16] Alpha = ['0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'];
+
+    union FloatBits
+    {
+        T floatValue;
+        ulong ulongValue;
+    }
+
+    FloatBits fb;
+    fb.floatValue = val;
+    ulong ival = fb.ulongValue;
+
+    static if (is(T == float))
+    {
+        ulong mnt = ival & ((1L << 23) - 1);
+        int exp = (ival >> 23) & ((1L << 8) - 1);
+        int maxexp = 255;
+        string sgn = (ival >> (23 + 8)) & 1 ? "-" : "";
+        enum long bias = 127;
+    }
+    else
+    {
+        version (DigitalMars)
+        {
+            // hack to work around issue 20363
+            ival ^= rm;
+            ival ^= rm;
+        }
+
+        ulong mnt = ival & ((1L << 52) - 1);
+        int exp = (ival >> 52) & ((1L << 11) - 1);
+        int maxexp = 2047;
+        string sgn = (ival >> (52 + 11)) & 1 ? "-" : "";
+        enum long bias = 1023;
```

Comment:
You should be able to deduplicate this branch (modulo the DigitalMars hack) by using the floating point's mantissa and exponent properties, something like
```
ulong mnt = ival & ((1L << T.mant_dig) - 1);
int exp = (ival >> T.mant_dig) & ((1L << (T.sizeof * 8) - 1 -T.mant_dig) - 1);
enum maxexp = T.max_exp;
string sgn = (ival >> ((T.sizeof * 8) - 1)) & 1 ? "-" : "";
enum long bias = (maxexp+1)>>1;
```

Reply:
I'll try, to do so. :-)

Reply:
Thanks, its much easier knowing what the magic numbers mean.

Reply:
Changed this too.